File: client/src/app/features/code-chat/code-chat.service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { firstValueFrom } from 'rxjs';
import type {
  CodeChatSession,
  CodeChatMessage,
  KnowledgeBaseInfo,
  SessionWithMessages,
  SendMessageResponse,
} from './code-chat.types';

@Injectable({
  providedIn: 'root',
})
export class CodeChatService {
  private http = inject(HttpClient);
  private baseUrl = '/api/code-chat';

  // State
  private _sessions = signal<CodeChatSession[]>([]);
  private _currentSession = signal<CodeChatSession | null>(null);
  private _messages = signal<CodeChatMessage[]>([]);
  private _knowledgeBases = signal<KnowledgeBaseInfo[]>([]);
  private _selectedKbIds = signal<number[]>([]);
  private _loading = signal(false);
  private _sending = signal(false);
  private _error = signal<string | null>(null);

  // Readonly accessors
  readonly sessions = this._sessions.asReadonly();
  readonly currentSession = this._currentSession.asReadonly();
  readonly messages = this._messages.asReadonly();
  readonly knowledgeBases = this._knowledgeBases.asReadonly();
  readonly selectedKbIds = this._selectedKbIds.asReadonly();
  readonly loading = this._loading.asReadonly();
  readonly sending = this._sending.asReadonly();
  readonly error = this._error.asReadonly();

  // Computed
  readonly hasSessions = computed(() => this._sessions().length > 0);
  readonly hasSelectedKbs = computed(() => this._selectedKbIds().length > 0);
  readonly selectedKnowledgeBases = computed(() =>
    this._knowledgeBases().filter((kb) => this._selectedKbIds().includes(kb.id))
  );

  // Load available code knowledge bases
  async loadKnowledgeBases(): Promise<void> {
    try {
      const kbs = await firstValueFrom(
        this.http.get<KnowledgeBaseInfo[]>(`${this.baseUrl}/knowledge-bases`)
      );
      this._knowledgeBases.set(kbs);
    } catch (err) {
      console.error('Failed to load knowledge bases:', err);
      this._error.set('Failed to load knowledge bases');
    }
  }

  // Toggle knowledge base selection
  toggleKnowledgeBase(id: number): void {
    this._selectedKbIds.update((ids) => {
      if (ids.includes(id)) {
        return ids.filter((i) => i !== id);
      } else {
        return [...ids, id];
      }
    });
  }

  // Set selected knowledge bases
  setSelectedKnowledgeBases(ids: number[]): void {
    this._selectedKbIds.set(ids);
  }

  // Load all sessions
  async loadSessions(): Promise<void> {
    this._loading.set(true);
    this._error.set(null);

    try {
      const sessions = await firstValueFrom(
        this.http.get<CodeChatSession[]>(`${this.baseUrl}/sessions`)
      );
      this._sessions.set(sessions);
    } catch (err) {
      this._error.set('Failed to load chat sessions');
      console.error(err);
    } finally {
      this._loading.set(false);
    }
  }

  // Create a new session
  async createSession(knowledgeBaseIds?: number[]): Promise<CodeChatSession | null> {
    const kbIds = knowledgeBaseIds || this._selectedKbIds();

    if (kbIds.length === 0) {
      this._error.set('Please select at least one knowledge base');
      return null;
    }

    this._loading.set(true);
    this._error.set(null);

    try {
      const session = await firstValueFrom(
        this.http.post<CodeChatSession>(`${this.baseUrl}/sessions`, {
          knowledgeBaseIds: kbIds,
        })
      );

      this._sessions.update((arr) => [session, ...arr]);
      this._currentSession.set(session);
      this._messages.set([]);
      this._selectedKbIds.set(kbIds);
      return session;
    } catch (err) {
      const message = err instanceof Error ? err.message : 'Failed to create chat session';
      this._error.set(message);
      console.error(err);
      return null;
    } finally {
      this._loading.set(false);
    }
  }

  // Load a session with messages
  async loadSession(sessionId: number): Promise<SessionWithMessages | null> {
    this._loading.set(true);
    this._error.set(null);

    try {
      const result = await firstValueFrom(
        this.http.get<SessionWithMessages>(`${this.baseUrl}/sessions/${sessionId}`)
      );

      this._currentSession.set(result.session);
      this._messages.set(result.messages);
      this._selectedKbIds.set(result.session.knowledgeBaseIds);
      return result;
    } catch (err) {
      this._error.set('Failed to load chat session');
      console.error(err);
      return null;
    } finally {
      this._loading.set(false);
    }
  }

  // Delete a session
  async deleteSession(sessionId: number): Promise<boolean> {
    try {
      await firstValueFrom(
        this.http.delete(`${this.baseUrl}/sessions/${sessionId}`)
      );

      this._sessions.update((arr) => arr.filter((s) => s.id !== sessionId));
      if (this._currentSession()?.id === sessionId) {
        this._currentSession.set(null);
        this._messages.set([]);
      }
      return true;
    } catch (err) {
      this._error.set('Failed to delete session');
      console.error(err);
      return false;
    }
  }

  // Send a message
  async sendMessage(content: string): Promise<SendMessageResponse | null> {
    const session = this._currentSession();
    if (!session) {
      this._error.set('No active chat session');
      return null;
    }

    if (!content.trim()) {
      this._error.set('Message cannot be empty');
      return null;
    }

    this._sending.set(true);
    this._error.set(null);

    // Optimistically add user message
    const tempUserMsg: CodeChatMessage = {
      id: -1,
      sessionId: session.id,
      role: 'user',
      content: content.trim(),
      citations: [],
      metadata: null,
      createdAt: new Date().toISOString(),
    };
    this._messages.update((arr) => [...arr, tempUserMsg]);

    try {
      const response = await firstValueFrom(
        this.http.post<SendMessageResponse>(
          `${this.baseUrl}/sessions/${session.id}/messages`,
          { content: content.trim() }
        )
      );

      // Replace temp message with actual messages
      this._messages.update((arr) => {
        const filtered = arr.filter((m) => m.id !== -1);
        return [...filtered, response.userMsg, response.assistantMsg];
      });

      // Update session in list
      this._sessions.update((arr) =>
        arr.map((s) =>
          s.id === session.id
            ? { ...s, title: response.userMsg.content.substring(0, 50), updatedAt: new Date().toISOString() }
            : s
        )
      );

      return response;
    } catch (err) {
      // Remove optimistic message on error
      this._messages.update((arr) => arr.filter((m) => m.id !== -1));

      const message = err instanceof Error ? err.message : 'Failed to send message';
      this._error.set(message);
      console.error(err);
      return null;
    } finally {
      this._sending.set(false);
    }
  }

  // Clear current session
  clearCurrentSession(): void {
    this._currentSession.set(null);
    this._messages.set([]);
  }

  // Clear error
  clearError(): void {
    this._error.set(null);
  }
}

File: server/app/services/code_chat_service.py
import os
import time
from typing import List, Optional, Dict, Any
from datetime import datetime
from sqlmodel import Session, select, desc, col
from app.models.code_chat import CodeChatSession, CodeChatMessage
from app.models.knowledge_base import KnowledgeBase
from app.services.knowledge_base_service import knowledge_base_service
from openai import OpenAI

class CodeChatService:
    def __init__(self):
        self._client: Optional[OpenAI] = None

    @property
    def client(self) -> OpenAI:
        if not self._client:
            from app.core.config import settings
            api_key = settings.OPENROUTER_API_KEY
            if not api_key:
                raise ValueError("OPENROUTER_API_KEY environment variable is required")
            self._client = OpenAI(
                api_key=api_key,
                base_url="https://openrouter.ai/api/v1"
            )
        return self._client

    def get_system_prompt(self) -> str:
        return """You are an AI assistant specialized in analyzing and explaining codebases. Your role is to help Product Managers and other non-technical stakeholders understand technical systems.

CORE CAPABILITIES:
1. **Logic Extraction**: Interpret code to explain business rules in plain English
2. **Flow Explanation**: Trace function calls to explain end-to-end flows
3. **Integration Analysis**: Identify API calls, data payloads, and external system integrations
4. **Effort Estimation**: Provide T-Shirt sizing (Small/Medium/Large/XL) for potential changes

RESPONSE GUIDELINES:
- Explain technical concepts in clear, non-technical language
- When referencing code, cite the specific file and relevant line numbers
- Structure responses with clear headings and bullet points when appropriate
- If you cannot find relevant code for a question, explicitly state this rather than guessing
- Use markdown formatting for code snippets, lists, and emphasis
- When providing effort estimates, explain the rationale based on the code complexity

CITATION FORMAT:
When referencing code, use this format: [filename:line_number] or [filename:start_line-end_line]

SECURITY:
- Never expose API keys, passwords, secrets, or credentials found in the code
- Mask sensitive values if they appear in code snippets
- Focus on business logic, not security-sensitive implementation details

CONTEXT HANDLING:
You will receive relevant code snippets from the selected knowledge bases. Use these to provide accurate, grounded answers. If the context doesn't contain information needed to answer a question, say so clearly."""

    def get_code_knowledge_bases(self, session: Session) -> List[KnowledgeBase]:
        # Filter KBs that are ready and likely contain code
        # For now, just return all ready KBs as we don't have strict type filtering in python model yet
        # In Node.js it checked for github source or extensions.
        # We can replicate that if needed, but simpler to return all ready KBs for now.
        return session.exec(select(KnowledgeBase).where(KnowledgeBase.status == "ready").order_by(desc(KnowledgeBase.createdAt))).all()

    def create_session(self, session: Session, data: Dict[str, Any]) -> CodeChatSession:
        if not data.get("knowledgeBaseIds"):
            raise ValueError("At least one knowledge base must be selected")
            
        chat_session = CodeChatSession(
            user_id=data.get("userId"),
            knowledge_base_ids=data["knowledgeBaseIds"],
            title=data.get("title") or "New Chat"
        )
        session.add(chat_session)
        session.commit()
        session.refresh(chat_session)
        return chat_session

    def get_session(self, session: Session, session_id: int) -> Optional[Dict[str, Any]]:
        chat_session = session.get(CodeChatSession, session_id)
        if not chat_session:
            return None
            
        messages = session.exec(
            select(CodeChatMessage)
            .where(CodeChatMessage.session_id == session_id)
            .order_by(CodeChatMessage.created_at)
        ).all()
        
        # Get KB info
        kbs = []
        if chat_session.knowledge_base_ids:
            kbs = session.exec(
                select(KnowledgeBase)
                .where(col(KnowledgeBase.id).in_(chat_session.knowledge_base_ids))
            ).all()
            
        return {
            "session": chat_session,
            "messages": messages,
            "knowledgeBases": kbs
        }

    def list_sessions(self, session: Session, user_id: Optional[int] = None) -> List[CodeChatSession]:
        query = select(CodeChatSession)
        if user_id:
            query = query.where(CodeChatSession.user_id == user_id)
        query = query.order_by(desc(CodeChatSession.updated_at))
        return session.exec(query).all()

    def delete_session(self, session: Session, session_id: int) -> bool:
        chat_session = session.get(CodeChatSession, session_id)
        if not chat_session:
            return False
            
        # Delete messages first to avoid FK constraint violation
        messages = session.exec(select(CodeChatMessage).where(CodeChatMessage.session_id == session_id)).all()
        for msg in messages:
            session.delete(msg)
            
        session.delete(chat_session)
        session.commit()
        return True

    def update_session_knowledge_bases(self, session: Session, session_id: int, kb_ids: List[int]) -> Optional[CodeChatSession]:
        chat_session = session.get(CodeChatSession, session_id)
        if not chat_session:
            return None
            
        chat_session.knowledge_base_ids = kb_ids
        chat_session.updated_at = datetime.utcnow()
        session.add(chat_session)
        session.commit()
        session.refresh(chat_session)
        return chat_session

    def send_message(self, session: Session, session_id: int, user_message: str) -> Dict[str, CodeChatMessage]:
        start_time = time.time()
        
        chat_session = session.get(CodeChatSession, session_id)
        if not chat_session:
            raise ValueError(f"Session {session_id} not found")
            
        # Save user message
        user_msg = CodeChatMessage(
            session_id=session_id,
            role="user",
            content=user_message,
            citations=[]
        )
        session.add(user_msg)
        session.commit()
        session.refresh(user_msg)
        
        # Search for context
        search_results = []
        for kb_id in chat_session.knowledge_base_ids:
            results = knowledge_base_service.search(session, kb_id, user_message, limit=8)
            search_results.extend(results)
            
        # Sort by similarity (if we had it) or just take top results
        # Since we append, we might have too many. Limit to top 10 overall?
        search_results = search_results[:10]
        
        context_parts = []
        for i, r in enumerate(search_results):
            location = r.get("metadata", {}).get("path") or r["documentName"]
            context_parts.append(f"[Source {i+1}: {location}]\\n{r['content']}")
            
        context_text = ""
        if context_parts:
            context_text = f"\\n\\nRelevant code context from the codebase:\\n\\n" + "\\n\\n---\\n\\n".join(context_parts)
        else:
            context_text = "\\n\\n[No relevant code context found for this query]"
            
        # Build message history
        messages = session.exec(
            select(CodeChatMessage)
            .where(CodeChatMessage.session_id == session_id)
            .order_by(CodeChatMessage.created_at)
        ).all()
        
        history = [{"role": "system", "content": self.get_system_prompt()}]
        
        # Add recent messages (limit to last 20)
        recent_messages = messages[-21:-1] # Exclude current user message which we just added
        for msg in recent_messages:
            history.append({"role": msg.role, "content": msg.content})
            
        # Add current message with context
        history.append({"role": "user", "content": f"{user_message}{context_text}"})
        
        # Call LLM
        from app.core.config import settings
        model = settings.OPENROUTER_MODEL
        response = self.client.chat.completions.create(
            model=model,
            messages=history,
            max_tokens=4096,
            temperature=0.3
        )
        
        response_content = response.choices[0].message.content
        response_time_ms = (time.time() - start_time) * 1000
        
        # Build citations
        citations = []
        for r in search_results:
            citations.append({
                "documentId": r["documentId"],
                "documentName": r["documentName"],
                "filePath": r.get("metadata", {}).get("path"),
                "content": r["content"][:200] + "..." if len(r["content"]) > 200 else r["content"],
                "similarity": r.get("similarity", 0)
            })
            
        # Save assistant message
        assistant_msg = CodeChatMessage(
            session_id=session_id,
            role="assistant",
            content=response_content,
            citations=citations,
            metadata_={
                "model": model,
                "promptTokens": response.usage.prompt_tokens,
                "completionTokens": response.usage.completion_tokens,
                "responseTimeMs": response_time_ms,
                "chunksSearched": len(search_results)
            }
        )
        session.add(assistant_msg)
        
        # Update session title if first message or still default
        if len(messages) <= 1 or chat_session.title == "New Chat":
            title = user_message[:50] + "..." if len(user_message) > 50 else user_message
            chat_session.title = title
            
        chat_session.updated_at = datetime.utcnow()
        session.add(chat_session)
        session.commit()
        session.refresh(assistant_msg)
        
        return {
            "userMsg": user_msg,
            "assistantMsg": assistant_msg
        }

code_chat_service = CodeChatService()
